README
NAME
    Package::Locator - Find a package among CPAN-like repositories

VERSION
    version 0.009

SYNOPSIS
      use Package::Locator;

      # Basic search...
      my $locator = Package::Locator->new();
      my $url = locator->locate( package => 'Test::More' );

      # Search for first within multiple repositories:
      my $repos = [ qw(http://cpan.pair.com http://my.company.com/DPAN) ];
      my $locator = Package::Locator->new( repository_urls => $repos );
      my $url = locator->locate( package => 'Test::More' );

      # Search for first where version >= 0.34:
      my $repos = [ qw(http://cpan.pair.com http://my.company.com/DPAN) ];
      my $locator = Package::Locator->new( repository_urls => $repos );
      my $url = locator->locate( package => 'Test::More' version => 0.34);

      # Search for latest where version  >= 0.34:
      my $repos = [ qw(http://cpan.pair.com http://my.company.com/DPAN) ];
      my $locator = Package::Locator->new( repository_urls => $repos );
      my $url = locator->locate( package => 'Test::More' version => 0.34, latest => 1);

      # Search for specific dist on multiple repositories...:
      my $repos = [ qw(http://cpan.pair.com http://my.company.com/DPAN) ];
      my $locator = Package::Locator->new( repository_urls => $repos );
      my $url = locator->locate( distribution => 'A/AU/AUTHOR/Foo-1.0.tar.gz');

DESCRIPTION
    Package::Locator attempts to answer the question: "Where can I find a
    distribution that will provide this package?" The answer is divined by
    searching the indexes for one or more CPAN-like repositories. If you
    also provide a specific version number, Package::Locator will attempt to
    find a distribution with that version of the package, or higher. You can
    also ask to find the latest version of a package across all the indexes.

    Package::Locator only looks at the index files for each repository, and
    those indexes only contain information about the latest versions of the
    packages within that repository. So Package::Locator is not BackPAN
    magic -- you cannot use it to find precisely which distribution a
    particular package (or file) came from. For that stuff, see "/See Also".

CONSTRUCTOR
  new( %attributes )
    All the attributes listed below can be passed to the constructor, and
    retrieved via accessor methods with the same name. All attributes are
    read-only, and cannot be changed once the object is constructed.

ATTRIBUTES
  repository_urls => [ qw(http://somewhere http://somewhere.else) ]
    An array reference containing the base URLs of the repositories you want
    to search. These are usually CPAN mirrors, but can be any website or
    local directory that is organized in a CPAN-like structure. For each
    request, repositories are searched in the order you specified them here.
    This defaults to http://cpan.perl.org.

  user_agent => $user_agent_obj
    The LWP::UserAgent object that will fetch index files. If you do not
    provide a user agent, then a default one will be constructed for you.

  cache_dir => '/some/directory/path'
    The path (as a string or Path::Class::Dir object) to a directory where
    the index file will be cached. If the directory does not exist, it will
    be created for you. If you do not specify a cache directory, then a
    temporary directory will be used. The temporary directory will be
    deleted when your application terminates.

  force => $boolean
    Causes any cached index files to be removed, thus forcing a new one to
    be downloaded when the object is constructed. This only has effect if
    you specified the "cache_dir" attribute. The default is false.

METHODS
  indexes()
    Returns a list of Package::Locator::Index objects representing the
    indexes of each of the repositories. The indexes are only populated
    on-demand when the "locate" method is called. The order of the indexes
    is the same as the order of the repositories defined by the
    "repository_urls" attribute.

  locate( package => 'Foo::Bar' )
  locate( package => 'Foo::Bar', latest => 1 )
  locate( package => 'Foo::Bar', version => '1.2')
  locate( package => 'Foo::Bar', version => '1.2', latest => 1 )
  locate ( distribution => 'A/AU/AUTHOR/Foo-Bar-1.0.tar.gz' )
    Given the name of a package, searches all the repository indexes and
    returns the URL to a distribution containing that requested package, or
    the distribution you requested.

    If you also specify a "version", then you'll always get a distribution
    that contains that version of the package or higher. If you also specify
    "latest" then you'll always get the distribution that contains the
    latest version of the package that can be found in all the indexes.
    Otherwise you'll just get the first distribution we can find that
    satisfies your request.

    If you give a distribution path instead, then you'll just get back the
    URL to the first distribution we find at that path in any of the
    repository indexes.

    If neither the package nor the distribution path can be found in any of
    the indexes, returns undef.

  clear_cache()
    Deletes the cached index files. Any subsequent calls to the "locate"
    method will cause the index files to be fetched anew.

MOTIVATION
    The CPAN module also provides a mechanism for locating packages or
    distributions, much like Package::Locator does. However, CPAN assumes
    that all repositories are CPAN mirrors, so it only searches the first
    repository that it can contact.

    My secret ambition is to fill the world with lots of DarkPAN
    repositories -- each with its own set of distributions. For that
    scenario, I need to search multiple repositories at the same time.

SEE ALSO
    If you need to locate a distribution that contains a precise version of
    a file rather than just a version that is "new enough", then look at
    some of these:

    Dist::Surveyor

    BackPAN::Index

    BackPAN::Version::Discover

SUPPORT
  Perldoc
    You can find documentation for this module with the perldoc command.

      perldoc Package::Locator

  Websites
    The following websites have more information about this module, and may
    be of help to you. As always, in addition to those websites please use
    your favorite search engine to discover more resources.

    *   Search CPAN

        The default CPAN search engine, useful to view POD in HTML format.

        <http://search.cpan.org/dist/Package-Locator>

    *   CPAN Ratings

        The CPAN Ratings is a website that allows community ratings and
        reviews of Perl modules.

        <http://cpanratings.perl.org/d/Package-Locator>

    *   CPAN Testers

        The CPAN Testers is a network of smokers who run automated tests on
        uploaded CPAN distributions.

        <http://www.cpantesters.org/distro/P/Package-Locator>

    *   CPAN Testers Matrix

        The CPAN Testers Matrix is a website that provides a visual overview
        of the test results for a distribution on various Perls/platforms.

        <http://matrix.cpantesters.org/?dist=Package-Locator>

    *   CPAN Testers Dependencies

        The CPAN Testers Dependencies is a website that shows a chart of the
        test results of all dependencies for a distribution.

        <http://deps.cpantesters.org/?module=Package::Locator>

  Bugs / Feature Requests
    <https://github.com/thaljef/Package-Locator/issues>

  Source Code
    <https://github.com/thaljef/Package-Locator>

      git clone git://github.com/thaljef/Package-Locator.git

AUTHOR
    Jeffrey Ryan Thalhammer <jeff@imaginative-software.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Imaginative Software Systems.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
0.009     2013-06-06 19:23:13 America/Los_Angeles

    Fixed some test failures reported by CPAN Testers.  
    But no code changes in the library itself.

0.008     2013-06-06 14:04:54 America/Los_Angeles

    Also use IO::Zlib instead of PerlIO::gzip during tests.

0.007     2013-06-05 22:18:56 America/Los_Angeles

    Now uses IO::Zlib instead of PerlIO::gzip so there is one less
    non-core dependency.

0.006     2012-05-04 16:00:24 America/Los_Angeles

    Added clear_cache() method on the Package::Locator class.
    This deletes the cached index files, causing subsequent
    calls to locate() to fetch the index files anew.

    When calling locate() with a distribution path, we actually
    try and fetch the corresponding URL at each repository.
    This allows you to find distributions that are not in
    the index (i.e. older relases).

0.004     2011-12-02 03:12:06 America/Los_Angeles

    Package::Locator::Index does its own parsing of the index file now,
    and can return the contents as a simple data structure.  Yes,
    I reinvented that wheel.  I have my reasons.

    The locate() method now uses named arguments.  The old positional
    argument interface is not supported.

    The "get_latest" attribute is gone.  Instead, you can specify a
    "latest" argument to the locate() method.

    Fixed some failing tests.  Thanks CPANTESTERS.

0.003     2011-11-23 14:02:56 America/Los_Angeles

    Fixed bug that inadvertenty swallowed all exceptions
    when calling the locate() method.

    Now requires Try::Tiny.

0.002     2011-11-21 21:50:22 America/Los_Angeles

    Found a bunch of bugs by writing test cases.  The first release
    didn't really work as advertised.  This one actually has a chance.

0.001     2011-11-21 15:32:09 America/Los_Angeles

    Initial release.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Find a package among CPAN-like repositories",
  "dist_author" => [
    "Jeffrey Ryan Thalhammer <jeff\@imaginative-software.com>"
  ],
  "dist_name" => "Package-Locator",
  "dist_version" => "0.009",
  "license" => "perl",
  "module_name" => "Package::Locator",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "CPAN::DistnameInfo" => 0,
    "Carp" => 0,
    "File::Temp" => 0,
    "IO::Zlib" => 0,
    "LWP::UserAgent" => 0,
    "Moose" => 0,
    "MooseX::MarkAsMethods" => 0,
    "MooseX::Types::Path::Class" => 0,
    "MooseX::Types::URI" => 0,
    "Path::Class" => 0,
    "URI" => 0,
    "URI::Escape" => 0,
    "perl" => "5.006",
    "version" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Find" => 0,
    "FindBin" => 0,
    "Test::Exception" => 0,
    "Test::More" => 0,
    "strict" => 0,
    "warnings" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Package/Locator/Index.pm
# ABSTRACT: The package index of a repository

package Package::Locator::Index;

use Moose;
use MooseX::Types::URI qw(Uri);
use MooseX::Types::Path::Class;
use MooseX::MarkAsMethods (autoclean => 1);

use Carp;
use File::Temp;
use Path::Class;
use IO::Zlib;
use LWP::UserAgent;
use URI::Escape;
use URI;

#------------------------------------------------------------------------

our $VERSION = '0.009'; # VERSION

#------------------------------------------------------------------------


has repository_url => (
    is        => 'ro',
    isa       => Uri,
    required  => 1,
    coerce    => 1,
);

#------------------------------------------------------------------------


has user_agent => (
   is          => 'ro',
   isa         => 'LWP::UserAgent',
   default     => sub { LWP::UserAgent->new() },
);

#------------------------------------------------------------------------


has cache_dir => (
   is         => 'ro',
   isa        => 'Path::Class::Dir',
   default    => sub { Path::Class::Dir->new( File::Temp::tempdir(CLEANUP => 1) ) },
   coerce     => 1,
);

#------------------------------------------------------------------------


has force => (
   is         => 'ro',
   isa        => 'Bool',
   default    => 0,
);

#------------------------------------------------------------------------


has index_file => (
    is         => 'ro',
    isa        => 'Path::Class::File',
    init_arg   => undef,
    lazy_build => 1,
);

#------------------------------------------------------------------------


has distributions => (
   is         => 'ro',
   isa        => 'HashRef',
   init_arg   => undef,
   default    => sub { $_[0]->_data->{distributions} },
   lazy       => 1,
);


#------------------------------------------------------------------------



has packages => (
   is         => 'ro',
   isa        => 'HashRef',
   init_arg   => undef,
   default    => sub { $_[0]->_data->{packages} },
   lazy       => 1,
);


#------------------------------------------------------------------------


has _data => (
    is         => 'ro',
    isa        => 'HashRef',
    init_arg   => undef,
    lazy_build => 1,
);

#------------------------------------------------------------------------------

sub _build_index_file {
    my ($self) = @_;

    my $repos_url = $self->repository_url->canonical()->as_string();
    $repos_url =~ s{ /*$ }{}mx;         # Remove trailing slash
    $repos_url = URI->new($repos_url);  # Reconstitute as URI object

    my $cache_dir = $self->cache_dir->subdir( URI::Escape::uri_escape($repos_url) );
    $self->__mkpath($cache_dir);

    my $destination = $cache_dir->file('02packages.details.txt.gz');
    $destination->remove() if -e $destination and $self->force();

    my $source = URI->new( "$repos_url/modules/02packages.details.txt.gz" );

    my $response = $self->user_agent->mirror($source, $destination);
    $self->__handle_ua_response($response, $source, $destination);

    return $destination;
}

#------------------------------------------------------------------------------

sub _build__data {
    my ($self) = @_;

    my $file = $self->index_file->stringify;
    my $fh = IO::Zlib->new($file, 'rb') or croak "Failed to open index file $file: $!";
    my $index_data = $self->__read_index($fh);
    close $fh;

    return $index_data;
}

#------------------------------------------------------------------------------

sub __read_index {
    my ($self, $fh) = @_;

    my $inheader      = 1;
    my $packages      = {};
    my $distributions = {};
    my $source        = $self->repository_url();

    while (<$fh>) {

        if ($inheader) {
            $inheader = 0 if not m/ \S /x;
            next;
        }

        chomp;
        my ($package, $version, $dist_path) = split;
        my $dist_struct = $distributions->{$dist_path} ||= { source => $source, path => $dist_path };
        my $pkg_struct  = {name => $package, version => $version, distribution => $dist_path};
        push @{ $dist_struct->{packages} ||= [] }, $pkg_struct;
        $packages->{$package} = $pkg_struct;

    }

    return { packages      => $packages,
             distributions => $distributions };
}

#------------------------------------------------------------------------

sub __handle_ua_response {
    my ($self, $response, $source, $destination) = @_;

    return 1 if $response->is_success();   # Ok
    return 1 if $response->code() == 304;  # Not modified
    croak sprintf 'Request to %s failed: %s', $source, $response->status_line();
}

#------------------------------------------------------------------------------

sub __mkpath {
    my ($self, $dir) = @_;

    return 1 if -e $dir;
    $dir = dir($dir) unless eval { $dir->isa('Path::Class::Dir') };
    return $dir->mkpath() or croak "Failed to make directory $dir: $!";
}

#------------------------------------------------------------------------

__PACKAGE__->meta->make_immutable();

#------------------------------------------------------------------------
1;

__END__

=pod

=for :stopwords Jeffrey Ryan Thalhammer Imaginative Software Systems

=head1 NAME

Package::Locator::Index - The package index of a repository

=head1 VERSION

version 0.009

=head1 SYNOPSIS

  use Package::Locator::Index;

  my $index = Package::Locator::Index->new( repository_url => 'http://somewhere' );
  my $dist  = $index->distributions->{'A/AU/AUTHOR/Foo-Bar-1.0.tar.gz'};
  my $pkg   = $index->packages->{'Foo::Bar'};

=head1 DESCRIPTION

B<This is a private module and there are no user-serviceable parts
here.  The API documentation is for my own reference only.>

L<Package::Locator::Index> is yet-another module for parsing the
contents of the F<02packages.details.txt> file from a CPAN-like
repository.

=head1 CONSTRUCTOR

=head2 new( %attributes )

All the attributes listed below can be passed to the constructor, and
can be retrieved via accessor methods with the same name.  All
attributes are read-only, and cannot be changed once the object is
constructed.

=head1 ATTRIBUTES

=head2 repository_url => 'http://somewhere'

The base URL of the repository you want to get the index from.  This
is usually a CPAN mirror, but can be any site or directory that is
organized in a CPAN-like structure.  This attribute is required.

=head2 user_agent => $user_agent_obj

The L<LWP::UserAgent> object that will fetch the index file.  If you
do not provide a user agent, then a default one will be constructed
for you.

=head2 cache_dir => '/some/directory/path'

The path (as a string or L<Path::Class::Dir> object) to a directory
where the index file will be cached.  If the directory does not exist,
it will be created for you.  If you do not specify a cache directory,
then a temporary directory will be used.  The temporary directory will
be deleted when your application terminates.

=head2 force => $boolean

Causes any cached index files to be removed, thus forcing a new one to
be downloaded when the object is constructed.  This only has effect if
you specified the C<cache_dir> attribute.  The default is false.

=head1 METHODS

=head2 index_file()

Returns the path to the local copy of the index file (as a
L<Path::Class::File>).

=head2 distributions

Returns a hashref representing the contents of the index.  The keys
are the paths to the distributions (as they appear in the index).  The
values are data structures that look like this:

  {
    path     => 'A/AU/AUTHOR/FooBar-1.0.tar.gz',
    source   => 'http://some.cpan.mirror'
    packages => [ ## See package structure below ## ]
  }

=head2 packages

Returns a hashref representing the contents of the index.  The keys
are the names of packages.  The values are data structures that look
like this:

  {
    name         => 'Foo',
    version      => '1.0',
    distribution => 'A/AU/AUTHOR/FooBar-1.0.tar.gz'
  }

=head1 MOTIVATION

There are numerous existing modules for parsing the
F<02packages.details.txt> file, but I wasn't completely happy with any
of them.  Most of the existing modules transform the data into various
flavors of Distribution and Package objects. But I'm not ready to
commit to any particular API for Distributions and Packages (not even
one of my own).  So L<Package::Locator::Index> exposes the index data
as simple data structures.

=head1 AUTHOR

Jeffrey Ryan Thalhammer <jeff@imaginative-software.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Imaginative Software Systems.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Package/Locator.pm
# ABSTRACT: Find a package among CPAN-like repositories

package Package::Locator;

use Moose;
use MooseX::MarkAsMethods (autoclean => 1);
use MooseX::Types::Path::Class;

use Carp;
use File::Temp;
use Path::Class;
use LWP::UserAgent;
use CPAN::DistnameInfo;
use URI;

use Package::Locator::Index;

use version;


#------------------------------------------------------------------------------

our $VERSION = '0.009'; # VERSION

#------------------------------------------------------------------------------


has repository_urls => (
    is         => 'ro',
    isa        => 'ArrayRef[URI]',
    auto_deref => 1,
    default    => sub { [URI->new('http://cpan.perl.org')] },
);

#------------------------------------------------------------------------------


has user_agent => (
   is          => 'ro',
   isa         => 'LWP::UserAgent',
   default     => sub { LWP::UserAgent->new() },
);

#------------------------------------------------------------------------------


has cache_dir => (
   is         => 'ro',
   isa        => 'Path::Class::Dir',
   default    => sub { Path::Class::Dir->new( File::Temp::tempdir(CLEANUP => 1) ) },
   coerce     => 1,
);

#------------------------------------------------------------------------------


has force => (
   is         => 'ro',
   isa        => 'Bool',
   default    => 0,
);

#------------------------------------------------------------------------------


has indexes => (
   is         => 'ro',
   isa        => 'ArrayRef[Package::Locator::Index]',
   auto_deref => 1,
   lazy_build => 1,
   init_arg   => undef,
);


#------------------------------------------------------------------------------

sub _build_indexes {
    my ($self) = @_;

    my @indexes = map { Package::Locator::Index->new( force          => $self->force(),
                                                      cache_dir      => $self->cache_dir(),
                                                      user_agent     => $self->user_agent(),
                                                      repository_url => $_ )
    } $self->repository_urls();

    return \@indexes;
}

#------------------------------------------------------------------------------


sub locate {
    my ($self, %args) = @_;

    $self->_validate_locate_args(%args);

    if ($args{package}) {
        my $package = $args{package};
        my $version = $args{version} || 0;
        my $latest  = $args{latest}  || 0;
        return $self->_locate_package($package, $version, $latest);
    }
    else {
        my $dist = $args{distribution};
        return $self->_locate_dist($dist);
    }
}

#------------------------------------------------------------------------------

sub _validate_locate_args {
    my ($self, %args) = @_;

    croak 'Cannot specify latest and distribution together'
        if $args{distribution} and $args{latest};

    croak 'Cannot specify version and distribution together'
        if $args{distribution} and $args{version};

    croak 'Cannot specify package and distribution together'
        if $args{distribution} and $args{package};

    croak 'Must specify package or distribution'
        if not ( $args{distribution} or $args{package} );

    return 1;
}

#------------------------------------------------------------------------------

sub _locate_package {
    my ($self, $package, $version, $latest) = @_;

    my $wanted_version = version->parse($version);

    my ($latest_found_package, $found_in_index);
    for my $index ( $self->indexes() ) {

        my $found_package = $index->packages->{$package};
        next if not $found_package;

        my $found_package_version = version->parse( $found_package->{version} );
        next if $found_package_version < $wanted_version;

        $found_in_index       ||= $index;
        $latest_found_package ||= $found_package;
        last unless $latest;

        ($found_in_index, $latest_found_package) = ($index, $found_package)
            if $self->__compare_packages($found_package, $latest_found_package) == 1;
    }


    if ($latest_found_package) {
        my $base_url = $found_in_index->repository_url();
        my $latest_dist_path = $latest_found_package->{distribution};
        return  URI->new( "$base_url/authors/id/" . $latest_dist_path );
    }

    return;
}

#------------------------------------------------------------------------------

sub _locate_dist {
    my ($self, $dist_path) = @_;

    for my $index ( $self->indexes ) {
        my $base_url = $index->repository_url();
        my $dist_url =  URI->new("$base_url/authors/id/$dist_path");

        return $dist_url if $index->distributions->{$dist_path};
        return $dist_url if $self->user_agent->head($dist_url)->is_success;
    }


    return;
}

#------------------------------------------------------------------------------

sub __compare_packages {
    my ($self, $pkg_a, $pkg_b) = @_;

    my $pkg_a_version = $self->__versionize( $pkg_a->{version} );
    my $pkg_b_version = $self->__versionize( $pkg_b->{version} );

    # TODO: compare dist mtimes (but they are on the server!)
    return  $pkg_a_version  <=> $pkg_b_version;
}

#------------------------------------------------------------------------------

sub __versionize {
    my ($self, $version) = @_;

    my $v = eval { version->parse($version) };

    return defined $v ? $v : version->new(0);
}

#------------------------------------------------------------------------------


sub clear_cache {
    my ($self) = @_;

    for my $index ( $self->indexes() ) {
        $index->index_file->remove();
    }

    $self->clear_indexes();

    return $self;
}

#------------------------------------------------------------------------------

__PACKAGE__->meta->make_immutable();

#------------------------------------------------------------------------------

1;

__END__

=pod

=for :stopwords Jeffrey Ryan Thalhammer Imaginative Software Systems cpan testmatrix url
annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata
placeholders metacpan

=head1 NAME

Package::Locator - Find a package among CPAN-like repositories

=head1 VERSION

version 0.009

=head1 SYNOPSIS

  use Package::Locator;

  # Basic search...
  my $locator = Package::Locator->new();
  my $url = locator->locate( package => 'Test::More' );

  # Search for first within multiple repositories:
  my $repos = [ qw(http://cpan.pair.com http://my.company.com/DPAN) ];
  my $locator = Package::Locator->new( repository_urls => $repos );
  my $url = locator->locate( package => 'Test::More' );

  # Search for first where version >= 0.34:
  my $repos = [ qw(http://cpan.pair.com http://my.company.com/DPAN) ];
  my $locator = Package::Locator->new( repository_urls => $repos );
  my $url = locator->locate( package => 'Test::More' version => 0.34);

  # Search for latest where version  >= 0.34:
  my $repos = [ qw(http://cpan.pair.com http://my.company.com/DPAN) ];
  my $locator = Package::Locator->new( repository_urls => $repos );
  my $url = locator->locate( package => 'Test::More' version => 0.34, latest => 1);

  # Search for specific dist on multiple repositories...:
  my $repos = [ qw(http://cpan.pair.com http://my.company.com/DPAN) ];
  my $locator = Package::Locator->new( repository_urls => $repos );
  my $url = locator->locate( distribution => 'A/AU/AUTHOR/Foo-1.0.tar.gz');

=head1 DESCRIPTION

L<Package::Locator> attempts to answer the question: "Where can I find
a distribution that will provide this package?"  The answer is divined
by searching the indexes for one or more CPAN-like repositories.  If
you also provide a specific version number, L<Package::Locator> will
attempt to find a distribution with that version of the package, or
higher.  You can also ask to find the latest version of a package
across all the indexes.

L<Package::Locator> only looks at the index files for each repository,
and those indexes only contain information about the latest versions
of the packages within that repository.  So L<Package::Locator> is not
BackPAN magic -- you cannot use it to find precisely which
distribution a particular package (or file) came from.  For that
stuff, see C<"/See Also">.

=head1 CONSTRUCTOR

=head2 new( %attributes )

All the attributes listed below can be passed to the constructor, and
retrieved via accessor methods with the same name.  All attributes are
read-only, and cannot be changed once the object is constructed.

=head1 ATTRIBUTES

=head2 repository_urls => [ qw(http://somewhere http://somewhere.else) ]

An array reference containing the base URLs of the repositories you
want to search.  These are usually CPAN mirrors, but can be any
website or local directory that is organized in a CPAN-like structure.
For each request, repositories are searched in the order you specified
them here.  This defaults to http://cpan.perl.org.

=head2 user_agent => $user_agent_obj

The L<LWP::UserAgent> object that will fetch index files.  If you do
not provide a user agent, then a default one will be constructed for
you.

=head2 cache_dir => '/some/directory/path'

The path (as a string or L<Path::Class::Dir> object) to a directory
where the index file will be cached.  If the directory does not exist,
it will be created for you.  If you do not specify a cache directory,
then a temporary directory will be used.  The temporary directory will
be deleted when your application terminates.

=head2 force => $boolean

Causes any cached index files to be removed, thus forcing a new one to
be downloaded when the object is constructed.  This only has effect if
you specified the C<cache_dir> attribute.  The default is false.

=head1 METHODS

=head2 indexes()

Returns a list of L<Package::Locator::Index> objects representing the
indexes of each of the repositories.  The indexes are only populated
on-demand when the C<locate> method is called.  The order of the
indexes is the same as the order of the repositories defined by the
C<repository_urls> attribute.

=head2 locate( package => 'Foo::Bar' )

=head2 locate( package => 'Foo::Bar', latest => 1 )

=head2 locate( package => 'Foo::Bar', version => '1.2')

=head2 locate( package => 'Foo::Bar', version => '1.2', latest => 1 )

=head2 locate ( distribution => 'A/AU/AUTHOR/Foo-Bar-1.0.tar.gz' )

Given the name of a package, searches all the repository indexes and
returns the URL to a distribution containing that requested package,
or the distribution you requested.

If you also specify a C<version>, then you'll always get a
distribution that contains that version of the package or higher.  If
you also specify C<latest> then you'll always get the distribution
that contains the latest version of the package that can be found in
all the indexes.  Otherwise you'll just get the first distribution we
can find that satisfies your request.

If you give a distribution path instead, then you'll just get back the
URL to the first distribution we find at that path in any of the
repository indexes.

If neither the package nor the distribution path can be found in any
of the indexes, returns undef.

=head2 clear_cache()

Deletes the cached index files.  Any subsequent calls to the C<locate>
method will cause the index files to be fetched anew.

=head1 MOTIVATION

The L<CPAN> module also provides a mechanism for locating packages or
distributions, much like L<Package::Locator> does.  However, L<CPAN>
assumes that all repositories are CPAN mirrors, so it only searches
the first repository that it can contact.

My secret ambition is to fill the world with lots of DarkPAN
repositories -- each with its own set of distributions.  For that
scenario, I need to search multiple repositories at the same time.

=head1 SEE ALSO

If you need to locate a distribution that contains a precise version
of a file rather than just a version that is "new enough", then look
at some of these:

L<Dist::Surveyor>

L<BackPAN::Index>

L<BackPAN::Version::Discover>

=head1 SUPPORT

=head2 Perldoc

You can find documentation for this module with the perldoc command.

  perldoc Package::Locator

=head2 Websites

The following websites have more information about this module, and may be of help to you. As always,
in addition to those websites please use your favorite search engine to discover more resources.

=over 4

=item *

Search CPAN

The default CPAN search engine, useful to view POD in HTML format.

L<http://search.cpan.org/dist/Package-Locator>

=item *

CPAN Ratings

The CPAN Ratings is a website that allows community ratings and reviews of Perl modules.

L<http://cpanratings.perl.org/d/Package-Locator>

=item *

CPAN Testers

The CPAN Testers is a network of smokers who run automated tests on uploaded CPAN distributions.

L<http://www.cpantesters.org/distro/P/Package-Locator>

=item *

CPAN Testers Matrix

The CPAN Testers Matrix is a website that provides a visual overview of the test results for a distribution on various Perls/platforms.

L<http://matrix.cpantesters.org/?dist=Package-Locator>

=item *

CPAN Testers Dependencies

The CPAN Testers Dependencies is a website that shows a chart of the test results of all dependencies for a distribution.

L<http://deps.cpantesters.org/?module=Package::Locator>

=back

=head2 Bugs / Feature Requests

L<https://github.com/thaljef/Package-Locator/issues>

=head2 Source Code


L<https://github.com/thaljef/Package-Locator>

  git clone git://github.com/thaljef/Package-Locator.git

=head1 AUTHOR

Jeffrey Ryan Thalhammer <jeff@imaginative-software.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Imaginative Software Systems.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
